art-decisions/proposals/2022-11-20-arcaflow-signals.md

```diff
@@ -0,0 +1,209 @@
+# Adding signals to the Arcaflow execution model
+
+## Background
+
+On Friday the 17th of November 2022 we had a very lengthy discussion with the Red Hat Performance & Scale Middleware
+team. They have introduced us to their extremely powerful tool for executing workflows called
+[qDup](https://github.com/Hyperfoil/qDup). This tool not only allows people to write workflows, but also dynamically
+react to outputs of programs and publish signals on an internal messaging bus. Workflow parts can communicate with each
+other even while running, sending and waiting for signals.
+
```

Comment:
Another aspect of qDup signals that we did not discuss was that they are in effect global, resettable countdown latches. This provides a number of key capabilities for qDup scripts, such as; 
- a waiting script can wait until for `1..n` other scripts to raise a signal before proceeding
- when signals have reached a counter of 0, they will no longer block scripts in the future (i.e. a condition has been met) 
- signals are resettable, so the pre-conditions for proceeding can be reset (useful for loops)
- there are specific qDup commands that allow for looping and branching based on signal states, e.g. `repeat-until` signal state is reached, `read-signal` for branching depending on signal state, `wait-for` signal to be raised etc